SNOW-1631831: JDBC - handle TCP connection closure (traffic dependency request)

sfc-gh-jszczerbinski opened the pull request:
# Overview

SNOW-1631831

## Pre-review self checklist
- [ ] PR branch is updated with all the changes from `master` branch
- [ ] The code is correctly formatted (run `mvn -P check-style validate`)
- [ ] New public API is not unnecessary exposed (run `mvn verify` and inspect `target/japicmp/japicmp.html`)
- [ ] The pull request name is prefixed with `SNOW-XXXX: `
- [ ] Code is in compliance with internal logging requirements

## External contributors - please answer these questions before submitting a pull request. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Issue: #NNNN


2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency or upgrading an existing one
   - [ ] I am adding new public/protected component not marked with `@SnowflakeJdbcInternalApi` (note that public/protected methods/fields in classes marked with this annotation are already internal)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

sfc-gh-mhofman commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockTest.java:
```diff
@@ -0,0 +1,250 @@
+                                waitForWiremock();
+                                return true;
+                            } catch (Exception e) {
+                                logger.warn("Failed to start wiremock, retrying: ", e);
```
I'm wondering maybe that one day for some reason it may hang forever. So maybe a retry max count would be nice to have and when reached this exception rethrown, makes sense?

sfc-gh-mhofman commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockTest.java:
```diff
@@ -0,0 +1,250 @@
+            request.setEntity(new StringEntity(mapping));
+            CloseableHttpResponse response = httpClient.execute(request);
+            if (response.getStatusLine().getStatusCode() != 200) {
+                System.out.println(response.getStatusLine().getStatusCode());
```
why not to use logger?

sfc-gh-mhofman commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockTest.java:
```diff
@@ -0,0 +1,250 @@
+            }
+            assumeTrue( response.getStatusLine().getStatusCode() == 200);
+        } catch (Exception e) {
+            e.printStackTrace();
```
and here also I would use logger and said maybe that setting wiremock mapping failed so it is more self explanatory in comparison to plain stack trace

sfc-gh-mhofman commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockTest.java:
```diff
@@ -0,0 +1,250 @@
+import static org.junit.Assume.*;
+
+public class RestRequestWiremockTest {
+    private static final SFLogger logger = SFLoggerFactory.getLogger(ProxyLatestIT.class);
```
fix class passed to the logger factory

sfc-gh-mhofman commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockTest.java:
```diff
@@ -0,0 +1,250 @@
+    @BeforeClass
+    public static void setUpClass() {
+        assumeFalse(RunningNotOnJava8.isRunningOnJava8());
+        assumeFalse(RunningNotOnJava21.isRunningOnJava21());
```
Why do we need to skip running those tests on those version?

sfc-gh-jszczerbinski commented:
- Moved the wiremock test functionality to BaseWiremockTest (so ProxyLatestIT and RestRequestWiremockTest can share functionality)
- Fixed the review comments

sfc-gh-mkubik commented on src/test/java/net/snowflake/client/jdbc/RestRequestWiremockIT.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import net.snowflake.client.category.TestCategoryOthers;
+import net.snowflake.client.core.ExecTimeTelemetryData;
```
Do we know why there's some failing test regarding this line?
```
 Compilation failure: Compilation failure: 
Error:  /mnt/host/src/test/java/net/snowflake/client/jdbc/RestRequestWiremockIT.java:[5,33] cannot find symbol
```

sfc-gh-jszczerbinski replied:
The name of the test influences how it's build and run. Changing name to one that ends in LatestIT fixed the issue (this test was 'inspired by' ProxyLatestIT)
